Frameworks/Starboard/PathMapper.cpp

```diff
@@ -0,0 +1,136 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include "Starboard.h"
+#include "ErrorHandling.h"
+#include "StringHelpers.h"
+
+#include <algorithm>
+#include <list>
+#include <regex>
+
+#include "Platform/EbrPlatform.h"
+#include "PathMapper.h"
+
+// utility function to tokenize string using delmiters
+// d:\src/winobjc ==> d:, src, winobjc
+// /src/winobjc ==> "", src, winobjc
+// / ==> ""
+// src ==> "src"
+
+static const std::wstring c_currentDir(L".");
+static const std::wstring c_UpDir(L"..");
+
+static std::list<std::wstring> _TokenizeString(const std::wstring& str, const wchar_t* delims) {
+    std::list<std::wstring> components;
+
+    std::size_t start = 0; // start from 0
+    std::size_t delimPos = str.find_first_of(delims);
+
+    while (start != std::wstring::npos) {
+        components.emplace_back(str.substr(start, delimPos - start));
+        start = str.find_first_not_of(delims, delimPos);
+        delimPos = str.find_first_of(delims, start);
+    }
+    return components;
+}
+
+// converts any ? to +, we may add more later if needed for compat
+
+static void _EscapeIllegalPathCharacters(std::wstring& str) {
+    std::size_t pos = str.find_first_of(L'?');
+    while (pos != std::wstring::npos) {
+        str[pos] = L'+';
+        pos = str.find_first_of(L'?', pos + 1);
+    }
+}
+
+// normalize relative path
+//  1 remove any "." or "" components
+//  2 from reverse, remove any ".." and preceeding component
+// if components becomes empty, return "."
+
+static void _NormalizeRelativePathComponents(std::list<std::wstring>& components) {
+    for (auto it = components.begin(); it != components.end();) {
+        if (*it == c_currentDir || it->empty()) {
+            it = components.erase(it);
+            continue;
+        } else if (*it == c_UpDir) {
+            // remove previous component if it exists
+            if (it == components.begin()) {
+                // error, no previous directory exists
+                components.clear();
+                break;
+            }
+            // we have to do this in case the last component is ..
+            it = components.erase(--it);
```

Comment:
They can be double linked I thought, but given how we're using it might make more sense to just use a std::vector.  FWIU it's generally better to take a hit every once in a while for resizing if we need to than having to jump around when iterating.